### ISSUE TITLE
Issue with cython 0.23

### ISSUE THREAD

**Issue by andreabedini** [PyTables/PyTables#481]
We might have an issue with the new cython release.

Users are not affected because we ship pre-cythoned c extensions. To reproduce:

```
$ virtualenv venv
$ source venv/bin/activate
$ pip install cython
$ pip install https://github.com/PyTables/PyTables/archive/v.3.2.1.zip
```

Then

    python -m tables.tests.test_tables RecArrayRangeTestCase.test01a_range

gives (I'm showing only the first error)

```
======================================================================
ERROR: None (__main__.RecArrayRangeTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/andreabedini/venv/lib/python2.7/site-packages/tables/tests/test_tables.py", line 2169, in test01a_range
    self.check_range()
  File "/Users/andreabedini/venv/lib/python2.7/site-packages/tables/tests/test_tables.py", line 2042, in check_range
    recarray = table.read(self.start, self.stop, self.step)
  File "/Users/andreabedini/venv/lib/python2.7/site-packages/tables/table.py", line 1965, in read
    arr = self._read(start, stop, step, field, out)
  File "/Users/andreabedini/venv/lib/python2.7/site-packages/tables/table.py", line 1887, in _read
    self.row._fill_col(result, start, stop, step, field)
  File "tables/tableextension.pyx", line 1272, in tables.tableextension.Row._fill_col (tables/tableextension.c:15021)
ValueError: could not broadcast input array from shape (2) into shape (0)

----------------------------------------------------------------------
```

I managed to track this down to [tableextension.pyx#L1247](https://github.com/PyTables/PyTables/blob/develop/tables/tableextension.pyx#L1247)

```python
        # Compute the end for this iteration
        stopr = startr + ((istopb - istartb - 1) / istep)
```

With cython 0.22 it becomes

```c
__pyx_t_8 = ((__pyx_v_istopb - __pyx_v_istartb) - 1);
      if (unlikely(__pyx_v_istep == 0)) {
        #ifdef WITH_THREAD
        PyGILState_STATE __pyx_gilstate_save = PyGILState_Ensure();
        #endif
        PyErr_SetString(PyExc_ZeroDivisionError, "integer division or modulo by zero");
        #ifdef WITH_THREAD
        PyGILState_Release(__pyx_gilstate_save);
        #endif
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1219; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      else if (sizeof(PY_LONG_LONG) == sizeof(long) && (!(((PY_LONG_LONG)-1) > 0)) && unlikely(__pyx_v_istep == (PY_LONG_LONG)-1)  && unlikely(UNARY_NEG_WOULD_OVERFLOW(__pyx_t_8))) {
        #ifdef WITH_THREAD
        PyGILState_STATE __pyx_gilstate_save = PyGILState_Ensure();
        #endif
        PyErr_SetString(PyExc_OverflowError, "value too large to perform division");
        #ifdef WITH_THREAD
        PyGILState_Release(__pyx_gilstate_save);
        #endif
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1219; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_v_stopr = ((__pyx_v_startr + __Pyx_div_PY_LONG_LONG(__pyx_t_8, __pyx_v_istep)) + 1);
```

while with cython 0.23

```c
      __pyx_t_8 = ((__pyx_v_istopb - __pyx_v_istartb) - 1);
      if (unlikely(__pyx_v_istep == 0)) {
        PyErr_SetString(PyExc_ZeroDivisionError, "integer division or modulo by zero");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1253; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      else if (sizeof(hsize_t) == sizeof(long) && (!(((PY_LONG_LONG)-1) > 0)) && unlikely(__pyx_v_istep == (PY_LONG_LONG)-1)  && unlikely(UNARY_NEG_WOULD_OVERFLOW(__pyx_t_8))) {
        PyErr_SetString(PyExc_OverflowError, "value too large to perform division");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1253; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_v_stopr = (__pyx_v_startr + __Pyx_div_hsize_t(__pyx_t_8, __pyx_v_istep));
```

Notice the different call for the division (`__Pyx_div_hsize_t` vs `__Pyx_div_PY_LONG_LONG`) in particular I believe `__pyx_v_istep` gets coerced to unsigned in some way.

**Comment by scopatz** [comment]
Weird, maybe we need to be more explicit about the types.

**Comment by andreabedini** [comment]
@scopatz I agree it's a typing problem. I had to look it up but it seems that C dictates that `istopb - istartb - 1` is an unsigned expression (because `istartb` is unsigned and `istopb` is signed but not big enough to contain `istartb`).

**Comment by andreabedini** [comment]
Hopefully fixed by 44dba04, but don't trust me. I tried reading that function and I got lost following indices around.

**Comment by ziggi0703** [comment]
With your fix, using Cython==0.23 worked fine. I had no errors in the short test.
